### PR TITLE
If platform is not running, always return an anonymous logger

### PR DESCRIPTION
**Pull request by laeubi**
Currently a logger is returned (and cached) regardless of the platform
is (still) running, that is the OSGi trackers are open.

This now guards the access to loggers, check for null in cases where
still the service is not running anymore, clears the loggers on shutdown
and return an anonymous logger in cases where platform is shut down.

Fix https://github.com/eclipse-platform/eclipse.platform/issues/62

### CODE REVIEW

**Comment by iloveeclipse** [comment]
@laeubi : i'm not an OSGI or service expert, so don't expect any qualified review on this PR.
It sounds as it would work :-)
I home @tjwatson can review that.

File: runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/InternalPlatform.java [hunk 1]
```diff
@@ -745,6 +752,13 @@ private DebugOptions getDebugOptions() {
 	}
 
 	private void closeOSGITrackers() {
+		ExtendedLogReaderService logReader = logReaderTracker.getService();
+		logs.values().removeIf(log -> {
```
Comment (tjwatson) [inline runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/InternalPlatform.java:756]: Probably shouldn't even do this stream if `logReader` is null.

A stream with `removeIf` seems like an expensive way to do this.  Why not do this?
```java
ExtendedLogReaderService logReader = logReaderTracker.getService();
if (logReader != null) {
  logs.values().forEach((log) -> logReader.removeLogListener(log));
}
logs.clear();
```

Comment (laeubi) [reply runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/InternalPlatform.java:756]: > A stream with `removeIf` seems like an expensive way to do this. Why not do this?

Not sure if we really safe anything here:

1. logReader should be not null in almost all cases
2. iterate first and then clear seem not much less "expensive" as we are shut down and there are not millions of bundles having a logger anyways...

Comment (tjwatson) [reply runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/InternalPlatform.java:756]: This is clearly a style discussion.  I think only doing a single null check is better on its own, but I also think the 5 lines of code is more simple and doing a single `clear()` call is more optimal than removing each element individually.

Is it going to make any measurable difference, I doubt seriously it.  But I still think it is better.

Comment (laeubi) [reply runtime/bundles/org.eclipse.core.runtime/src/org/eclipse/core/internal/runtime/InternalPlatform.java]: Done!

**Review by tjwatson — APPROVED** [review]
LGTM

**Review by iloveeclipse — APPROVED** [review]

**Pull request merged**